<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>genotools</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">genotools</h1>



<p>The <strong>genotools</strong> package primarily provides functions
to help make and use pre-made polygenic scores with MoBa data in
p471.</p>
<p><em>This testing version of 0.3.1 implements a new pipeline using
ldpred2 as the primary software for making scores. Recommended QC of
summary statistics prior to estimating weights in ldpred2 is also added.
Most downstream functions (fetch_pgs, etc) are backwards compatible with
prsice2.</em></p>
<div id="using-pre-made-scores" class="section level2">
<h2>Using pre-made scores</h2>
<p>The first step is to ascertain which PGS are available and what names
you need to refer to them by in your request. This is done using the
<code>available_pgs</code> function. This can be used in conjunction
with <code>pgs_search</code>, a free text search function:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">available_pgs</span>() <span class="sc">%&gt;%</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a>  <span class="fu">pgs_search</span>(<span class="st">&quot;adhd&quot;</span>)</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="co">#&gt; Scanning PGS directory...</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="co">#&gt;   Pheno_shortname                                Phenotype Outcome_type Source</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="co">#&gt; 1        adhd2022 Attention deficit hyperactivity disorder       binary    PGC</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="co">#&gt;   Ref (PMID)                  Sumstats_filename</span></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="co">#&gt; 1   36702997 ADHD2022_iPSYCH_deCODE_PGC.meta.gz</span></span></code></pre></div>
<p>If you are conducting within-family analyses (for example, trio-pgs
models) it is recommended to use a more stringent filter on imputation
quality. Setting the option to <code>within_fam = TRUE</code> shows
scores that have been estimated using the MoBa genetic data that only
includes variants with INFO scores &gt;= 0.95.</p>
<p>The output provides all the details we have in the metadata,
including where the sumstats are sourced from. The Pheno_shortname
column has the unique shortened version of the phenotype name that can
be used to reference the scores in other functions.</p>
<p>To get scores from the directory for use in analyses, you need to run
the <code>fetch_pgs</code> function:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>adhd_mdd_pgs <span class="ot">&lt;-</span> <span class="fu">fetch_pgs</span>(<span class="fu">c</span>(<span class="st">&quot;adhd2022&quot;</span>,<span class="st">&quot;mdd2025&quot;</span>)) </span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="co">#&gt; Retrieving available PGS...</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="co">#&gt; Fetching adhd2022 scores... (This is #1 of 2 available phenotypes requested)</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="co">#&gt; Fetching mdd2025 scores... (This is #2 of 2 available phenotypes requested)</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="fu">head</span>(adhd_mdd_pgs <span class="sc">%&gt;%</span> </span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a>  <span class="fu">mask_ids</span>()) <span class="co">#We replace the actual IDs here for demo purposes</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a><span class="co">#&gt; # A tibble: 6 x 6</span></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a><span class="co">#&gt;   FID     IID     f_id    m_id    adhd2022_pgs mdd2025_pgs</span></span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a><span class="co">#&gt;   &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;          &lt;dbl&gt;       &lt;dbl&gt;</span></span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a><span class="co">#&gt; 1 0129317 1072765 0852686 0219341         1.73       0.963</span></span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a><span class="co">#&gt; 2 1716040 1992006 1415322 2015315         1.37       1.06 </span></span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a><span class="co">#&gt; 3 1831577 0918252 1262805 1563012         1.46       1.16 </span></span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a><span class="co">#&gt; 4 1502170 1428522 0322537 1327746         1.22       1.09 </span></span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a><span class="co">#&gt; 5 0788795 0607299 1011094 1092671         1.01       1.46 </span></span>
<span id="cb2-19"><a href="#cb2-19" tabindex="-1"></a><span class="co">#&gt; 6 1014591 1843019 1705216 0081578         1.42       0.859</span></span></code></pre></div>
<p>The fetched PGS dataframe has IID (individual ID, sentrix) and FID
(family id, plink/KING) as identification variables. For scores
estimated using LDpred2, these are from the auto model. Use
<code>within_fam = TRUE</code> to fetch scores for use in within family
analyses. Importantly, these scores <strong>are not</strong> controlled
for PCs, genotyping batch, or imputation batch. To do so you need to use
<code>process_pgs</code>.</p>
<p>See the help pages for prsice2 scores and their proptions regarding
maf, clumping, thresholds, and more.</p>
</div>
<div id="making-polygenic-scores" class="section level2">
<h2>Making polygenic scores</h2>
<p>If you need polygenic scores that are not in the shared directory,
the <code>make_pgs</code> function can, with correctly formatted summary
statistics and appropriate inputs, complete all steps necessary to
estimate weights and PGS using LDpred2. This involves preparing and
submitting a job script to the HPC, so a bit of set up is required.</p>
<div id="steps-to-install-r-packages-needed-to-create-pgs-using-ldpred2-on-colossus" class="section level3">
<h3>Steps to install R packages needed to create PGS using LDpred2 on
Colossus</h3>
<ol style="list-style-type: decimal">
<li><p>Open a Colossus session using putty</p></li>
<li><p>In the command line: <code>module load R/4.4.2-gfbf-2024a</code>
<code>R</code></p></li>
<li><p>R session should open; check that it is version 4.4.2. Install
the required packages using:
<code>install.packages(c(&quot;ggplot2&quot;,&quot;bigsnpr&quot;,&quot;bigreadr&quot;,&quot;data.table&quot;,&quot;tidyverse&quot;, &quot;Cairo&quot;))</code>.
Agree to using a local library.</p></li>
<li><p>After all packages are installed quit R session:<br />
`q()`` No need to save a workspace image.</p></li>
<li><p>Back in your Windows-vm R session, make sure you have the
<code>ssh</code> package installed to allow genotools to initialise a
connection to the cluster.</p></li>
</ol>
<p>Now you should be ready to create a score. The <code>make_pgs</code>
function makes a directory in you working directory with the name of you
tsd username (input for the user argument). In this directory will be
another directory with the shortname of the PGS. The function creates
and R script with some QC of the summary statistics and estimation of
the weights and scores for both the LDpred2 infinitesimal and auto model
as well as the submission script in this directory. The function will
then automatically submit the job to the cluster.</p>
<p>If do not want to submit the job automatically you can set the
argument <code>submit_ssh = FALSE</code>. You may want to do this if,
for example, you would like to make changes to the QC or want the QC
graphs to save (see below). Note if you manually submit the submission
script on the cluster, you must have your working directory be the
directory the submission and R script is in.</p>
<p>If you want to make scores for within family analyses set
<code>genotype_data = &quot;genoHapMap3plus_filter_N200k.rds&quot;</code></p>
<p>Study the documentation for the <code>make_pgs</code> function, and
familiarise yourself with the guidance for creating scores using LDpred2
(e.g., <a href="https://privefl.github.io/bigsnpr/articles/LDpred2.html" class="uri">https://privefl.github.io/bigsnpr/articles/LDpred2.html</a>)
to ensure you have correctly formatted summary statistics. While we have
tried to make the QC pipeline accurate and robust, it is ultimately your
responsibility to ensure everything has run correctly. Note that in some
cases scores can still be returned even when there are issues with the
QC, so it is important to check and read through output files and
perform sanity checks on the PGS. See below for an example of the
correct inputs:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="fu">library</span>(genotools)</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="fu">make_pgs</span>(<span class="at">user =</span> <span class="st">&quot;p471-lauriejh&quot;</span>,</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>         <span class="at">host=</span><span class="st">&quot;p471-hpc-01.tsd.usit.no&quot;</span>,</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>         <span class="at">software =</span> <span class="st">&quot;ldpred2&quot;</span>,</span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>         <span class="at">shortname =</span> <span class="st">&quot;diet-pc1&quot;</span>,</span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>         <span class="at">account=</span><span class="st">&quot;p471&quot;</span>,</span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a>         <span class="at">cpu_time=</span><span class="st">&quot;05:00:00&quot;</span>,</span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>         <span class="at">memory=</span><span class="st">&quot;2G&quot;</span>,</span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a>         <span class="at">cores=</span><span class="dv">32</span>,</span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a>         <span class="at">lib=</span><span class="st">&quot;//ess/p471/data/durable/common/rlibs/4-3-2&quot;</span>,</span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a>         <span class="at">sumstats_dir=</span><span class="st">&quot;//ess/p471/data/no-backup/gwas_sumstats/&quot;</span>,</span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a>         <span class="at">outputs_dir=</span><span class="st">&quot;//ess/p471/data/durable/people/laurie&quot;</span>,</span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a>         <span class="at">sumstats_filename =</span> <span class="st">&quot;BOLTlmm_UKB_genoQCEURN455146_v3_diet_ffq_median_PC1_BgenSnps_mac20_maf0.005_info0.6.gz&quot;</span>,</span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a>         <span class="at">genotype_dir=</span><span class="st">&quot;//ess/p471/data/durable/data/genetic/MoBaPsychGen_v1/&quot;</span>,</span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a>         <span class="at">genotype_data=</span><span class="st">&quot;genoHapMap3plus_N200k.rds&quot;</span>,</span>
<span id="cb3-18"><a href="#cb3-18" tabindex="-1"></a>         <span class="at">ldref_dir=</span><span class="st">&quot;//ess/p471/data/durable/data/genetic/ldref/&quot;</span>,</span>
<span id="cb3-19"><a href="#cb3-19" tabindex="-1"></a>         <span class="at">case_control =</span><span class="cn">FALSE</span>,</span>
<span id="cb3-20"><a href="#cb3-20" tabindex="-1"></a>         <span class="at">A1 =</span> <span class="st">&quot;ALLELE1&quot;</span>,</span>
<span id="cb3-21"><a href="#cb3-21" tabindex="-1"></a>         <span class="at">A2 =</span> <span class="st">&quot;ALLELE0&quot;</span>,</span>
<span id="cb3-22"><a href="#cb3-22" tabindex="-1"></a>         <span class="at">stat =</span> <span class="st">&quot;BETA&quot;</span>,</span>
<span id="cb3-23"><a href="#cb3-23" tabindex="-1"></a>         <span class="at">stat_type =</span> <span class="st">&quot;BETA&quot;</span>,</span>
<span id="cb3-24"><a href="#cb3-24" tabindex="-1"></a>         <span class="at">pvalue =</span> <span class="st">&quot;P_BOLT_LMM&quot;</span>,</span>
<span id="cb3-25"><a href="#cb3-25" tabindex="-1"></a>         <span class="at">SNP =</span> <span class="st">&quot;SNP&quot;</span>,</span>
<span id="cb3-26"><a href="#cb3-26" tabindex="-1"></a>         <span class="at">CHR=</span> <span class="st">&quot;CHR&quot;</span>,</span>
<span id="cb3-27"><a href="#cb3-27" tabindex="-1"></a>         <span class="at">BP =</span> <span class="st">&quot;BP&quot;</span>,</span>
<span id="cb3-28"><a href="#cb3-28" tabindex="-1"></a>         <span class="at">SE =</span> <span class="st">&quot;SE&quot;</span>,</span>
<span id="cb3-29"><a href="#cb3-29" tabindex="-1"></a>         <span class="at">N =</span> <span class="dv">449210</span>,</span>
<span id="cb3-30"><a href="#cb3-30" tabindex="-1"></a>         <span class="at">info =</span> <span class="st">&quot;INFO&quot;</span>,</span>
<span id="cb3-31"><a href="#cb3-31" tabindex="-1"></a>         <span class="at">MAF =</span> <span class="st">&quot;A1FREQ&quot;</span>,</span>
<span id="cb3-32"><a href="#cb3-32" tabindex="-1"></a>         <span class="at">maf_threshold =</span> <span class="fl">0.01</span>,</span>
<span id="cb3-33"><a href="#cb3-33" tabindex="-1"></a>         <span class="at">source =</span> <span class="st">&quot;http://www.kp4cd.org/dataset_downloads/t2d&quot;</span>,</span>
<span id="cb3-34"><a href="#cb3-34" tabindex="-1"></a>         <span class="at">pmid =</span><span class="dv">32193382</span>,</span>
<span id="cb3-35"><a href="#cb3-35" tabindex="-1"></a>         <span class="at">pgs_meta=</span>genotools<span class="sc">::</span>pgs_metadata)</span></code></pre></div>
<p><code>fetch_pgs</code> can be used with scores created from
<code>make_pgs</code> in the same way as if getting scores from the PGS
repository. The only difference is that you will need to specify the
pgs_directory. This will be outputs_dir/user. So to fetch the scores
created in the above example, you add the following argument to
<code>fetch_pgs</code> :
<code>pgs_directory = &quot;//ess/p471/data/durable/people/laurie/p471-lauriejh&quot;</code></p>
</div>
<div id="making-multiple-scores-at-once" class="section level3">
<h3>Making multiple scores at once</h3>
<p>If you need to make multiple scores, you can create a dataframe with
the column names being the arguments for <code>make_pgs</code> and each
row being the inputs for those arguments for each PGS, for example:</p>
<pre><code>#&gt; # A tibble: 6 x 25
#&gt;   shortname  sumstats_dir       sumstats_filename case_control A1    A2    stat 
#&gt;   &lt;chr&gt;      &lt;chr&gt;              &lt;chr&gt;             &lt;lgl&gt;        &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;
#&gt; 1 adhd2022   //ess/p471/data/d~ ADHD2022_iPSYCH_~ TRUE         A1    A2    OR   
#&gt; 2 asd2019    //ess/p471/data/d~ iPSYCH-PGC_ASD_2~ TRUE         A1    A2    OR   
#&gt; 3 Neu2018b   //ess/p471/data/d~ sumstats_neuroti~ FALSE        A1    A2    BETA 
#&gt; 4 EA2018     //ess/p471/data/d~ EA3_excl_23andMe~ FALSE        EA    OA    BETA 
#&gt; 5 intell2018 //ess/p471/data/d~ SavageJansen_201~ FALSE        A1    A2    stdB~
#&gt; 6 anx2024    //ess/p471/data/d~ daner_fullANX_v1~ TRUE         A1    A2    OR   
#&gt; # i 18 more variables: stat_type &lt;chr&gt;, pvalue &lt;chr&gt;, SNP &lt;chr&gt;, CHR &lt;chr&gt;,
#&gt; #   BP &lt;chr&gt;, SE &lt;chr&gt;, N &lt;chr&gt;, info &lt;chr&gt;, MAF &lt;chr&gt;, source &lt;chr&gt;,
#&gt; #   MoBa_included_in_study &lt;lgl&gt;, QC_reviewed &lt;lgl&gt;,
#&gt; #   sumstats_orginal_location &lt;chr&gt;, outputs_dir &lt;chr&gt;, user &lt;chr&gt;,
#&gt; #   cpu_time &lt;chr&gt;, memory &lt;chr&gt;, pmid &lt;chr&gt;</code></pre>
<p>And then use the following code:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="fu">library</span>(purrr)</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>safe_pgs <span class="ot">&lt;-</span> <span class="fu">safely</span>(make_pgs) <span class="co">#wraps make_pgs in safely so if one PGS fails it moves on to the next instead of failing completely</span></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>run_pgs <span class="sc">%&gt;%</span></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a>  <span class="fu">pwalk</span>(safe_pgs)</span></code></pre></div>
<p>For traits already in the PGS repository there is an excel
spreadsheet
(//ess01/p471/data/durable/common/pgs_directory/pgs/ldpred2/sumstats_key.xlsx)
in this format that can be used to remake scores from the repository.
For example, if you need different QC parameters than the defaults. Note
you will need to remove columns that are not arguments in
<code>make_pgs</code> and add columns for user specific information (i.e
user, outputs_dir, etc.)</p>
</div>
<div id="processing-pgs-for-use-with-moba-phenotypic-data" class="section level3">
<h3>Processing PGS for use with MoBa phenotypic data</h3>
<p>To use the PGS with MoBa phenotypic data, you need preg_id and
BARN_NR variables. These can be retrieved for your dataframe by running
the <code>process_pgs</code> function on it. The other important feature
of this function is to provide essential covariates (20PCs, genotyping
batch, and imputation batch) and/or regress the raw PGS on these
covariates:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>adhd_mdd_pgs_procd <span class="ot">&lt;-</span> <span class="fu">process_pgs</span>(adhd_mdd_pgs,</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>                                  <span class="at">regress_pgs =</span> <span class="cn">TRUE</span>,</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>                                  <span class="at">return_covs=</span><span class="cn">TRUE</span>)</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="co">#&gt; Regressing out genotyping batch, imputation batch, and 20PCs from all supplied scores...</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a><span class="co">#&gt; Processing complete.</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="co">#&gt; PGS variables with _res suffix are standardised residuals from the regression model.</span></span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a><span class="co">#&gt; You also now have ID variables (preg_id, F_ID, M_ID) for linkage to phenotypic data.</span></span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a><span class="co"># You get back raw and residualised scores and the covariates with these options</span></span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a><span class="fu">names</span>(adhd_mdd_pgs_procd)</span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a><span class="co">#&gt;  [1] &quot;IID&quot;              &quot;FID&quot;              &quot;preg_id&quot;          &quot;BARN_NR&quot;         </span></span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a><span class="co">#&gt;  [5] &quot;f_id&quot;             &quot;m_id&quot;             &quot;Role&quot;             &quot;adhd2022_pgs&quot;    </span></span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a><span class="co">#&gt;  [9] &quot;mdd2025_pgs&quot;      &quot;adhd2022_pgs_res&quot; &quot;mdd2025_pgs_res&quot;  &quot;PC1&quot;             </span></span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a><span class="co">#&gt; [13] &quot;PC2&quot;              &quot;PC3&quot;              &quot;PC4&quot;              &quot;PC5&quot;             </span></span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a><span class="co">#&gt; [17] &quot;PC6&quot;              &quot;PC7&quot;              &quot;PC8&quot;              &quot;PC9&quot;             </span></span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a><span class="co">#&gt; [21] &quot;PC10&quot;             &quot;PC11&quot;             &quot;PC12&quot;             &quot;PC13&quot;            </span></span>
<span id="cb6-18"><a href="#cb6-18" tabindex="-1"></a><span class="co">#&gt; [25] &quot;PC14&quot;             &quot;PC15&quot;             &quot;PC16&quot;             &quot;PC17&quot;            </span></span>
<span id="cb6-19"><a href="#cb6-19" tabindex="-1"></a><span class="co">#&gt; [29] &quot;PC18&quot;             &quot;PC19&quot;             &quot;PC20&quot;             &quot;genotyping_batch&quot;</span></span>
<span id="cb6-20"><a href="#cb6-20" tabindex="-1"></a><span class="co">#&gt; [33] &quot;imputation_batch&quot;</span></span>
<span id="cb6-21"><a href="#cb6-21" tabindex="-1"></a></span>
<span id="cb6-22"><a href="#cb6-22" tabindex="-1"></a><span class="fu">head</span>(adhd_mdd_pgs_procd <span class="sc">%&gt;%</span> </span>
<span id="cb6-23"><a href="#cb6-23" tabindex="-1"></a>  <span class="fu">mask_ids</span>()) <span class="co">#We replace the actual IDs here for demo purposes)</span></span>
<span id="cb6-24"><a href="#cb6-24" tabindex="-1"></a><span class="co">#&gt; # A tibble: 6 x 33</span></span>
<span id="cb6-25"><a href="#cb6-25" tabindex="-1"></a><span class="co">#&gt;   IID     FID     preg_id BARN_NR f_id    m_id    Role  adhd2022_pgs mdd2025_pgs</span></span>
<span id="cb6-26"><a href="#cb6-26" tabindex="-1"></a><span class="co">#&gt;   &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;        &lt;dbl&gt;       &lt;dbl&gt;</span></span>
<span id="cb6-27"><a href="#cb6-27" tabindex="-1"></a><span class="co">#&gt; 1 0923883 1994378 1257584 1       0393781 0075699 Child         1.73       0.963</span></span>
<span id="cb6-28"><a href="#cb6-28" tabindex="-1"></a><span class="co">#&gt; 2 0506808 0586895 1288603 &lt;NA&gt;    0416152 1366987 Fath~         1.37       1.06 </span></span>
<span id="cb6-29"><a href="#cb6-29" tabindex="-1"></a><span class="co">#&gt; 3 1300699 0096909 1434405 &lt;NA&gt;    0274520 1260617 Moth~         1.46       1.16 </span></span>
<span id="cb6-30"><a href="#cb6-30" tabindex="-1"></a><span class="co">#&gt; 4 1294279 0275389 1494686 &lt;NA&gt;    0031686 2070693 Moth~         1.22       1.09 </span></span>
<span id="cb6-31"><a href="#cb6-31" tabindex="-1"></a><span class="co">#&gt; 5 1053796 0353612 0045269 &lt;NA&gt;    1452975 0510302 Fath~         1.01       1.46 </span></span>
<span id="cb6-32"><a href="#cb6-32" tabindex="-1"></a><span class="co">#&gt; 6 1952659 1248641 1424757 1       0748403 1596118 Child         1.42       0.859</span></span>
<span id="cb6-33"><a href="#cb6-33" tabindex="-1"></a><span class="co">#&gt; # i 24 more variables: adhd2022_pgs_res &lt;dbl&gt;, mdd2025_pgs_res &lt;dbl&gt;,</span></span>
<span id="cb6-34"><a href="#cb6-34" tabindex="-1"></a><span class="co">#&gt; #   PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;, PC3 &lt;dbl&gt;, PC4 &lt;dbl&gt;, PC5 &lt;dbl&gt;, PC6 &lt;dbl&gt;,</span></span>
<span id="cb6-35"><a href="#cb6-35" tabindex="-1"></a><span class="co">#&gt; #   PC7 &lt;dbl&gt;, PC8 &lt;dbl&gt;, PC9 &lt;dbl&gt;, PC10 &lt;dbl&gt;, PC11 &lt;dbl&gt;, PC12 &lt;dbl&gt;,</span></span>
<span id="cb6-36"><a href="#cb6-36" tabindex="-1"></a><span class="co">#&gt; #   PC13 &lt;dbl&gt;, PC14 &lt;dbl&gt;, PC15 &lt;dbl&gt;, PC16 &lt;dbl&gt;, PC17 &lt;dbl&gt;, PC18 &lt;dbl&gt;,</span></span>
<span id="cb6-37"><a href="#cb6-37" tabindex="-1"></a><span class="co">#&gt; #   PC19 &lt;dbl&gt;, PC20 &lt;dbl&gt;, genotyping_batch &lt;chr&gt;, imputation_batch &lt;chr&gt;</span></span></code></pre></div>
<p>The “_res” variables should be ready-to-use PGS for analyses. You
should perform sanity checks to validate this.</p>
</div>
</div>
<div id="overview-of-qc" class="section level2">
<h2>Overview of QC</h2>
<div id="basic-qc" class="section level3">
<h3>Basic QC</h3>
<ol style="list-style-type: decimal">
<li><p>If MAF or allele FRQ columns are provided, variants under
specified frequency are removed (default and for scores in the
repository this is 0.01).</p></li>
<li><p>If Info scores are provided, variants with INFO &lt; 0.6 are
removed.</p></li>
<li><p>Variants are matched to the reference panel (we are currently
using one for 1,444,196 HapMap3+ variants based on European individuals
of UK biobank: <a href="https://figshare.com/articles/dataset/LD_reference_for_HapMap3_/21305061" class="uri">https://figshare.com/articles/dataset/LD_reference_for_HapMap3_/21305061</a>)
so only variants in common between summary statistics and the reference
panel are kept. If less than 0.1% are flipped these SNPs are dropped
(assumed to be an error).</p></li>
</ol>
</div>
<div id="qc-based-on-comparing-standard-deviation-sd-of-variants" class="section level3">
<h3>QC based on comparing standard deviation (SD) of variants</h3>
<p>The basis of this QC is comparing the SD of variants calculated in
two different ways.</p>
<ol style="list-style-type: decimal">
<li><p>SD based on allele frq is calculated (sd_ldref). The default and
what is used for scores in the repository is to calculate this using the
allele frq from the reference panel. This is what is done in example
code from the 2020 Privé, et al. paper (<a href="https://github.com/privefl/paper-ldpred2/blob/master/code/example-with-provided-ldref.R" class="uri">https://github.com/privefl/paper-ldpred2/blob/master/code/example-with-provided-ldref.R</a>),
it has also been recommended to calculate this based on allele frq
provided in summary statistics and what is done in the code for the 2022
Privé, Arbel, et al. paper (<a href="https://github.com/privefl/paper-misspec/tree/main/code/prepare-sumstats" class="uri">https://github.com/privefl/paper-misspec/tree/main/code/prepare-sumstats</a>).
To use allele frq provided in summary statistics set option
<code>sd_maf = TRUE</code></p></li>
<li><p>SD based on N,SE, and beta (and estimated sd of the phenotype for
continuous GWAS) is calculated (sd_ss). If INFO scores are provided the
sd calculation is adjusted for imputation quality of the variants using
these. This step is a new addition in genotools 0.3.1. If you’d like to
not perform this correction, either filter on INFO &lt; 0.6 (or any
other desired threshold) prior to running make_pgs and do not include
the INFO column in the arguments of the function or hash out lines 486 -
491 and 502-506 in the .pipl script and submit the job
manually.</p></li>
<li><p>The two SD estimates are compared and variants not meeting QC
filters are removed. For all summary statistics, QC1 in Privé, Arbel, et
al. (2022) is performed this includes the following thresholds: sd_ss
&lt; (0.7 * sd_ldref) OR sd_ss &gt; (sd_ldref + 0.1) OR sd_ss &lt; 0.1
OR sd_ldref &lt; 0.05. This is slightly more restrictive than the QC
originally recommend in Privé, et al. (2020) and what was used in
versions previous to 0.3.1. This orginal QC was the same but the first
threshold was sd_ss &lt; (0.5 * sd_ldref).</p></li>
</ol>
<p>If you’d like to use this QC threshold then you can change line 529
and 535 to <code>is_bad_qc2 &lt;- is_bad</code> in the .pipl script and
submit it manually.</p>
<p>If allele frequency is provided in the summary statistics, the
default is that a third threshold is used in addition to the previous is
used, this is QC2 in Privé, Arbel, et al. (2022). This threshold removes
variants with an allele frequency that differs by &gt;0.1 between the
summary statistics and the reference panel. If you do not want this step
you can change line 535 to <code>is_bad_qc2 &lt;- is_bad2</code> in the
.pipl script and submit it manually. This will mean QC1 is
performed.</p>
</div>
<div id="how-to-interpret-qc-and-imputation-of-neff" class="section level3">
<h3>How to interpret QC and Imputation of Neff</h3>
<p><strong>Please look through the output file and double check that the
columns are correct, the inputs are being interpreted as intended, the
QC steps you intended ran, and a sensible number of variants are
matching to the reference panel estimation, etc. </strong></p>
<p><strong>The job may not fail and still create PGS even if there is
something wrong with these or if there are a large number of variants
failing QC.</strong></p>
<p>In the output file there is a line that say “xx SNP are bad”. This is
the number of SNPs removed from the QC based on SD of the variants. When
the submission script is submitted to colossus manually a plot comparing
the two SDs is created.</p>
<p>As a rule of thumb, lots of SNPs removed means there is likely
misspecification of the Neff. If sd_ldref &gt; sd_ss, variants will fall
below / to the right of the center line (when sd_ldref = sd_ss) in the
SD QC plot. This means sd_ss is underestimated - typically Neff is too
large. Likewise if they are above the line, the Neff is likely
overestimated. If the GWAS used BOLT-LMM or SAIGE then the provided Neff
may be under (SAIGE) or over (BOLT-LMM) estimated. There is additional
correction that can be done (global power improvement) if BOLT_LMM was
used. If this is the case, see Privé, Arbel, et al. (2022) for more
details.</p>
<p>If more half of the variants fail QC the Neff is automatically
imputed. If this is the case, <strong>double check that all columns
(particularly the N column) are correct</strong>. Common issues may
be:</p>
<ol style="list-style-type: upper-alpha">
<li><p>that the Neff is actually half of the effective sample
size</p></li>
<li><p>the Neff has been estimated from overall case/controls rather
than the sum of Neff calculated in each cohort separately.</p></li>
</ol>
<p>If the GWAS is a meta analysis and minimum INFO scores (as opposed to
average) are reported it may be worth rerunning and not correcting the
SD estimates with INFO scores. Please note that, particularly for
continuous traits, be skeptical of the imputation and double check what
is being done with the most recent recommendations from the
literature.</p>
<p>Another plot that is generated is the path of the chains from the
auto model. You should check this plot to make sure the chains converged
(the points should “travel” to the blue line). See <a href="https://privefl.github.io/bigsnpr/articles/LDpred2.html" class="uri">https://privefl.github.io/bigsnpr/articles/LDpred2.html</a>
for an example. Filtering of bad chains is implemented automatically in
the script.</p>
<p>If you have questions on or spot errors/bugs in the QC or
interpretation please email: <a href="mailto:laura.elizabeth.hegemann@fhi.no" class="email">laura.elizabeth.hegemann@fhi.no</a></p>
<div id="sources" class="section level4">
<h4>Sources:</h4>
<p>The LDpred2 pipeline is adapted from Andre Allegrini see <a href="https://github.com/AndreAllegrini/LDpred2/tree/main" class="uri">https://github.com/AndreAllegrini/LDpred2/tree/main</a>,
which was based on recommendations outline here: <a href="https://privefl.github.io/bigsnpr/articles/LDpred2.html" class="uri">https://privefl.github.io/bigsnpr/articles/LDpred2.html</a></p>
<p><strong>Reference panel was download here</strong>: <a href="https://figshare.com/articles/dataset/LD_reference_for_HapMap3_/21305061" class="uri">https://figshare.com/articles/dataset/LD_reference_for_HapMap3_/21305061</a>.
Recommend in this paper (citation to use):</p>
<p>Privé, F., Albiñana, C., Arbel, J., Pasaniuc, B., &amp; Vilhjálmsson,
B. J. (2022). Inferring disease architecture and predictive ability with
LDpred2-auto. The American Journal of Human Genetics, 110(12),
2042-2055.</p>
<p><strong>Current QC recommendations (citation to use):</strong></p>
<p>Privé, F., Arbel, J., Aschard, H., &amp; Vilhjálmsson, B. J. (2022).
Identifying and correcting for misspecifications in GWAS summary
statistics and polygenic scores. Human Genetics and Genomics Advances,
3(4), 100136.</p>
<p><strong>LDpred2 software citation and original QC
recommendations:</strong></p>
<p>Privé, F., Arbel, J., &amp; Vilhjálmsson, B. J. (2020). LDpred2:
better, faster, stronger. Bioinformatics, 36(22-23), 5424-5431.</p>
<p><strong>Useful source for calculating Neff:</strong></p>
<p><a href="https://github.com/GenomicSEM/GenomicSEM/wiki/2.1-Calculating-Sum-of-Effective-Sample-Size-and-Preparing-GWAS-Summary-Statistics" class="uri">https://github.com/GenomicSEM/GenomicSEM/wiki/2.1-Calculating-Sum-of-Effective-Sample-Size-and-Preparing-GWAS-Summary-Statistics</a></p>
</div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
